gpiozero/tones.py

```diff
@@ -0,0 +1,200 @@
+        represent as "A4". One semi-tone above this would be "A#4" or
+        alternatively "Bb4". Unicode representations of sharp and flat are also
+        accepted.
+        """
+        if isinstance(note, bytes):
+            note = note.decode('ascii')
+        if isinstance(note, str):
+            match = Tone.regex.match(note)
+            if match:
+                octave = int(match.group('octave')) + 1
+                return cls.from_midi(
+                    Tone.tones.index(match.group('note')) +
+                    Tone.semitones[match.group('semi')] +
+                    octave * 12)
+        raise ValueError('invalid note specification: %r' % note)
+
+    @classmethod
+    def from_frequency(cls, freq):
+        """
+        Construct a :class:`Tone` from a frequency specified in `Hz`_ which
+        must be a positive floating-point value in the range 0 < freq <= 20000.
+
+        .. _Hz: https://en.wikipedia.org/wiki/Hertz
+        """
+        if 0 < freq <= 20000:
+            return super(Tone, cls).__new__(cls, freq)
+        raise ValueError('invalid frequency: %.2f' % freq)
+
+    @property
+    def frequency(self):
+        """
+        Return the frequency of the tone in `Hz`_.
+
+        .. _Hz: https://en.wikipedia.org/wiki/Hertz
+        """
+        return float(self)
+
+    @property
+    def midi(self):
+        """
+        Return the (nearest) MIDI note to the tone's frequency. This will be an
+        integer number in the range 0 to 127. If the frequency is outside the
+        range represented by MIDI notes (which is approximately 8Hz to 12.5KHz)
+        :exc:`ValueError` exception will be raised.
+        """
+        result = int(round(12 * log2(self.frequency / 440) + 69))
+        if 0 <= result < 128:
+            return result
+        raise ValueError('%f is outside the MIDI note range' % self.frequency)
+
+    @property
+    def note(self):
+        """
+        Return the (nearest) note to the tone's frequency. This will be a
+        string in the form accepted by :meth:`from_note`. If the frequency is
+        outside the range represented by this format ("A0" is approximately
+        27.5Hz, and "G9" is approximately 12.5Khz) a :exc:`ValueError`
+        exception will be raised.
+        """
+        offset = self.midi - 60  # self.midi - A4_midi + Tone.tones.index('A')
+        index = offset % 12      # offset % len(Tone.tones)
+        octave = 4 + offset // 12
+        if 0 <= octave <= 9:
+            return (
+                Tone.tones[index] +
+                ('#' if Tone.tones[index] == Tone.tones[index - 1] else '') +
+                str(octave)
+            )
+        raise ValueError('%f is outside the notation range' % self.frequency)
+
+    def up(self, n=1):
+        """
+        Return the :class:`Tone` *n* semi-tones above this frequency (*n*
+        defaults to 1).
+        """
+        return Tone.from_midi(self.midi + n)
```

Comment:
I have absolutely zero musical background, but again I'm wondering why this is being done by converting to and from `.midi` instead of just using `.frequency` ?  (at the very least, presumably it introduces quantisation steps which the user might not expect from reading the docstring?)

Reply:
You're absolutely right that it involves quantization, but then that's the basis of MIDI (and indeed musical notes) generally. Quick bit of musical background (which I was astonished to realize I remember the other day!):

Take a note, say concert A (A4) which is 440Hz. If you want the same note an octave up you double the frequency, so A5 is 880Hz. Conversely the same note an octave down is half the frequency, so A3 is 220Hz. An octave is divided into 12 semi-tones (think of a semi-tone as the difference between one key and the next key on a piano keyboard - that's including the black keys, BTW). So, if 12 semi-tones (one octave) increases/decreases the frequency by a power of 2, what's the frequency ratio between one note and the next (or prior)? It's the twelfth root of 2<sup>*</sup>. So to go up a note, you multiple the frequency by the twelfth root of 2, to go down a note you divide by the same. This explains why we're using the MIDI system here: it keeps it simple. Going up a note is +1, going down a note is -1 and you don't start getting into twelfth roots and negative powers :)

<sup>*</sup> Or at least that's what we've mostly standardized on these days - the so called "[Equal temperament system](https://en.wikipedia.org/wiki/Equal_temperament)". I'd avoid looking at that link if you're not into musical theory - it's a *major* rabbit hole! Though it is quite interesting hearing music played in different tuning systems, given that some classical pieces were written prior to its invention and thus if played today on an equal temperament tuning, aren't necessarily played "how they were written".

Reply:
Ahhh, I didn't realise it was so complex :laughing: :notes: 
My concern was that if a user had a frequency which was part-way between two midi notes, they might expect the `up` or `down` to shift their original frequency, and not jump to the midi frequency nearest to their original frequency first?
But it sounds like it's complicated enough already, so I won't ask to make it more complicated! :grin:

Reply:
It's a reasonable concern, but if they're playing with frequencies which aren't already a note they're either playing with "pure" frequencies (i.e. not concerned with the musical aspect at all - which is fine, Tone derives from float after all so it can just be treated mathematically as such), in which case I'd *guess* they're not going to be interested in methods dealing with semi-tones, *or* they're playing in different tuning systems - which is sufficiently complex that I've not included any support for it.

... yet ;)

(actually I wouldn't know where to start with that stuff - it's something I *used* to know a tiny bit about 20+ years ago having studied it at a very high level without much detail - but I'd need to do a huge amount of reading before confidently designing an API to support it)

Reply:
> if they're playing with frequencies which aren't already a note they're playing with "pure" frequencies, in which case I'd _guess_ they're not going to be interested in methods dealing with semi-tones

Fair point, I hadn't considered that. Told you I wasn't musical :laughing: